src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs

```diff
@@ -644,6 +644,88 @@ protected override void ApplyAnalyzerReferenceRemoved(ProjectId projectId, Analy
+            }
+
+            var mappedChanges = mappedChangesBuilder.ToImmutable();
+            if (mappedChanges.Any())
+            {
+                // Group the mapped text changes by file, then apply all mapped text changes for the file.
+                foreach (var changesForFile in mappedChanges.GroupBy(change => change.FileName))
+                {
+                    using var invisibleEditor = new InvisibleEditor(ServiceProvider.GlobalProvider, changesForFile.Key, GetHierarchy(projectChanges.ProjectId), needsSave: true, needsUndoDisabled: false);
+                    TextEditApplication.UpdateText(changesForFile.Select(change => change.TextChange).ToImmutableArray(), invisibleEditor.TextBuffer, EditOptions.None);
+                }
+            }
+
+            static Dictionary<DocumentId, (ImmutableArray<MappedSpanResult> MappedSpans, ImmutableArray<TextChange> OriginalTextChanges)> GetTextChangesAndMappedSpans(
+                ProjectChanges projectChanges, ImmutableArray<DocumentId> changedMappedDocuments)
+            {
+                return Task.Run(async () =>
+                {
+                    var mappedSpansPerDocument = new Dictionary<DocumentId, (ImmutableArray<MappedSpanResult> MappedSpans, ImmutableArray<TextChange> OriginalTextChanges)>();
+                    foreach (var changedDocumentId in changedMappedDocuments)
+                    {
+                        var oldDocument = projectChanges.OldProject.GetDocument(changedDocumentId)!;
+                        var newDocument = projectChanges.NewProject.GetDocument(changedDocumentId)!;
+
+                        var mappingService = newDocument.Services.GetService<ISpanMappingService>();
+                        var textChanges = (await newDocument.GetTextChangesAsync(oldDocument, CancellationToken.None).ConfigureAwait(false)).ToImmutableArray();
+                        var mappedSpanResults = await mappingService.MapSpansAsync(oldDocument, textChanges.Select(tc => tc.Span), CancellationToken.None).ConfigureAwait(false);
+
+                        mappedSpansPerDocument[changedDocumentId] = (mappedSpanResults, textChanges);
+                    }
+
+                    return mappedSpansPerDocument;
+                }).WaitAndGetResult(CancellationToken.None);
+            }
+
+            static void MapTextChanges(CodeAnalysis.Document newDocument, ImmutableArray<MappedSpanResult> mappedSpanResults, ImmutableArray<TextChange> originalTextChanges, ArrayBuilder<(string, TextChange)> builder)
+            {
+                Contract.ThrowIfFalse(mappedSpanResults.Length == originalTextChanges.Length);
+
+                for (var i = 0; i < mappedSpanResults.Length; i++)
+                {
+                    // Only include changes that could be mapped.
+                    var newText = originalTextChanges[i].NewText;
+                    if (!mappedSpanResults[i].IsDefault && newText != null)
```

Comment:
Why didn't we just filter these earlier on in the GetTextChangesAndMappedSpans call?